doc/melange.rst

```diff
@@ -0,0 +1,66 @@
+.. _melange_main:
+
+***********************************
+JavaScript Compilation With Melange
+***********************************
+
+`Melange <https://github.com/melange-re/melange>`_ is a compiler from OCaml to
+JavaScript. Unlike Js_of_ocaml, Melange works by translating OCaml compiler
+internal lambda representation to JS files. This allows to produce a single
+JavaScript file from each OCaml module. Melange can be installed with
+`Opam, Esy or Nix <https://github.com/melange-re/melange#installation>`_ package
+managers.
+
+Compiling to JS
+===============
+
+Dune has experimental support for building Melange libraries. To use it, you
+must declare the ``melange`` extension in your ``dune-project`` file:
+
+.. code:: scheme
+
+  (lang dune 3.5)
+  (using melange 0.1)
+
+Then, given this example:
+
+.. code:: bash
+
+   echo 'Js.log "hello from melange"' > foo.ml
+
+With the following ``dune`` file:
+
+.. code:: scheme
+
+  (library (name foo) (modes melange))
+
+One can then request the ``.js`` target:
+
+.. code:: bash
+
+   $ dune build .foo.objs/melange/foo.js
+   $ node _build/default/.foo.objs/melange/foo.js
+   hello from melange
+
+At the moment, executable targets are not supported, but using explicit
+targets inside libraries as shown above should enable similar results.
+
+Faster builds with ``subdir`` and ``dirs`` stanzas
+==================================================
+
+Melange libraries are commonly installed from the npm package repository,
+together with other JavaScript packages. To avoid having Dune inspect
+unnecessary folders in ``node_modules``, it is recommended to explicitly
+whitelist those folders that are relevant for Melange builds.
+
+This can be accomplished by combining :ref:`subdir` and :ref:`dune-subdirs`
+stanzas in a ``dune`` file, and co-locate this file together with the
+``node_modules`` folder. The :ref:`dune-vendored_dirs` stanza can be used as
+well to avoid warnings in Melange libraries to appear in the application build.
+
+.. code:: scheme
+
+  (subdir
+   node_modules
+   (vendored_dirs reason-react)
+   (dirs reason-react))
```

Comment:
is `(dirs)` actually needed if `(vendored_dirs)` is passed?

Reply:
ah got it, `vendored_dirs` is implicitly empty while `dirs` is implicitly everything. this is a confusing model...